PIPs/pip-27.md

```diff
@@ -0,0 +1,127 @@
+There is no restriction for the maximum value.
+
+The config should have a private method to calculate the `RetentionBlocks` based on `RetentionDays`.
+The `RetentionBlocks` are the number of blocks that should be kept and not pruned.
+Given that each day has almost 8640 blocks, the number of blocks to keep is `RetentionBlocks = RetentionDays * 8640`.
+
+### PruneBlock Function
+
+The `PruneBlock` function removes a block and all transactions inside the block from the database.
+It is a private function and can't be accessed from outside the store.
+It accepts a `Batch` pointer and a block height, and returns a boolean indicating whether
+a block at the given height exists, and an error if there is any.
+The `PruneBlock` function first tries to retrieve the block at the given height and decode it.
+Once it is decoded, it iterates over all transactions and updates the batch
+by deleting the associated keys from the database.
+
+### Pruning Database
+
+A Full Node can convert to a Pruned Node by pruning the database and removing old blocks.
+Pruning the database is a time and resource-consuming process and should not be performed while the node is running.
+We need to add a new command named `prune` to `pactus-daemon` to prune an offline node.
+In the attachment to this proposal, there is a batch file for Windows users that can be used for pruning the node.
+
+Pruning the database works as follows:
+
+1. It checks if the `.pruning` file exists inside the database folder.
+   If it doesn't exist, it creates a `.pruning` file and
+   sets the content of the file to `LastBlockHeight - RetentionBlocks + 1`.
+2. It reads the content of the `.pruning` file and sets the pruning height to the content of the file minus 1.
+3. If the pruning height is less than or equal to zero, it exits.
+4. It updates the `.pruning` file and sets the pruning height.
+5. It calls the `PruneBlock` function and commits the batch.
+6. It repeats from step 2.
+
+### Pruning on New Block
+
+Once a new block is committed, the following operations should be performed:
+
+1. If the node is a Full Node, no action is needed.
+2. If the node is a Pruned Node, call the `PruneBlocks` function
+   with the block number equal to `LastBlockHeight - RetentionBlocks`.
+
+This process ensures that for each new block added,
+one old block will be removed, keeping the store blocks up to the `RetentionBlocks` number.
+
+### Importing Data
+
+Once a new node is initialized and before starting to sync with the network,
+it can download and import pruned data from a centralized server.
+This helps a pruned node sync faster.
+
+### Import Data in Pactus Daemon
+
+We need to add a new command named `import` to `pactus-daemon`.
+The `import` command shows a list of available databases that can be imported from a centralized server.
+Once the user chooses a database, it downloads and extracts it into the working directory.
+If a database already exists, it should show an error before downloading.
```

Comment:
In this case, the pruned node is downloading only some part of the chain! not the whole chain from Genesis, so can we still trust the centralized server by verifying?

We are only verifying and seeing some parts of the chain, is there any way for manipulation by a centralized server?

And one more important issue, we are syncing some blocks at all, what about indexed data extracted from transactions? balances, stakes, and in future contracts states?

When we have a whole chain, we can retrieve states by reading whole blocks and indexing the changes that a transaction makes, such as bonding a new validator, sending a native coin, or a contract deploy.

But when we don't have the whole chain we need these data and we need to verify them as well.

Reply:
Let's continue this discussion on https://github.com/pactus-project/PIPs/discussions/134